streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java

```diff
@@ -265,23 +264,45 @@ private InMemoryKeyValueIterator(final Set<Bytes> keySet, final boolean forward)
 
         @Override
         public boolean hasNext() {
-            return iter.hasNext();
+            if (!open) {
+                throw new InvalidStateStoreException(String.format("Store %s has closed", name));
```

Comment:
Thanks @ableegoldman for your review ! This check of open iterator is taken from the old `DelegatingPeekingKeyValueIterator`. It's just to avoid querying the iterator once it is closed. I kept the same error message that existed before, but maybe we will have to change it, as you said it's the iterator which is closing.

Reply:
Interesting, I'd say that's probably a bug in the `DelegatingPeekingKeyValueIterator` error message then.

I certainly doesn't hurt to throw an exception if the iterator is used after being closed, though I wonder if  `InvalidStateStoreException` really makes sense here. Imo we should update both the error message and the exception type. I'd say just a simple IllegalStateException would fit here

Can you make this change both for the new iterator here and for the `DelegatingPeekingKeyValueIterator`?

Reply:
Btw can you also rename the iterator's `open` field to `iteratorOpen` or something that won't conflict with the `open` field of the outer class ie the store itself? Also it probably doesn't need to be volatile, pretty sure iterators are not supposed to be shared between threads

Reply:
Agreed ! I made the changes